File: CppDesignPattern/CppDesignPattern/SOLID_OpenClosed.cpp
//#define OCP
#ifdef OCP
#include <iostream>
#include <vector>
#include <fstream>

enum class Color { red, green, blue }; 
enum class Size {small, medium, large};

struct Product {
	std::string name;
	Color color;
	Size size;
};

//// wrong
//struct ProductFilter
//{
//	std::vector<Product*> by_color(std::vector<Product*> items, Color color)
//	{
//		std::vector<Product*> result;
//		for (auto& i : items)
//		{
//			if (i->color == color)
//				result.push_back(i);
//		}
//		return result;
//	}
//
//	std::vector<Product*> by_color_and_size(std::vector<Product*> items, Color color, Size size)
//	{
//		std::vector<Product*> result;
//		for (auto& i : items)
//		{
//			if (i->color == color && i->size == size)
//				result.push_back(i);
//		}
//		return result;
//	}
//};

//struct ProductFilter
//{
//	std::vector<Product*> by_color(std::vector<Product*> items, Color color)
//	{
//		std::vector<Product*> result;
//		for (auto& i : items)
//		{
//			if (i->color == color)
//				result.push_back(i);
//		}
//		return result;
//	}
//};


template <typename T>
struct Specification
{
	virtual bool is_satisfied(T* item) = 0;
};

template <typename T>
struct Filter
{
	virtual std::vector<T*> filter(std::vector<T*> items, Specification<T>& spec) = 0;
};

template <typename  T>
struct AndSpecification : Specification<T>
{
	Specification<T>& first;
	Specification<T>& second;

	AndSpecification(Specification<T>& first, Specification<T>& second)
		: first(first), second(second) {}

	bool is_satisfied(Product* item) override {
		return first.is_satisfied(item) && second.is_satisfied(item);
	}
};

template<typename T> 
AndSpecification<T>& operator&&(const Specification<T>& first, const Specification<T>& second)
{
	return { first, second };
}



struct BetterFilter : Filter<Product>
{
	std::vector<Product*> filter(std::vector<Product*> items, Specification<Product>& spec) override
	{
		std::vector<Product*> result;
		for (auto& item : items)
			if (spec.is_satisfied(item))
				result.push_back(item);

		return result;
	}
};

struct ColorSpecification : Specification<Product>
{
	Color color;

	ColorSpecification(Color color) : color(color) {}

	bool is_satisfied(Product* item) override {
		return item->color == color;
	}
};

struct SizeSpecification : Specification<Product>
{
	Size size;

	SizeSpecification(Size size) : size(size) {}

	bool is_satisfied(Product* item) override {
		return item->size == size;
	}
};


int main()
{
	Product apple{ "Apple", Color::green, Size::small };
	Product tree{ "Tree", Color::green, Size::large};
	Product house{ "House", Color::blue, Size::large };

	std::vector<Product*> items{ &apple, &tree, &house };

	/*ProductFilter pf;
	const std::vector<Product*>& green_thing = pf.by_color(items, Color::green);

	for (auto& g : green_thing)
	{
		std::cout << g->name << std::endl;
	}*/

	BetterFilter bf;
	ColorSpecification green(Color::green);
	for (auto& item : bf.filter(items, green))
	{
		std::cout << item->name << " " <<  "green"  << std::endl;
	}

	AndSpecification<Product> green_and_large = ColorSpecification(Color::green) && SizeSpecification(Size::large);

	std::cin.get();

	return 0;
}

#endif

File: CSharpDesignPattern/CSharpDesignPattern/SOLID_OpenClosed.cs
#define OCP
#if OCP

using System;
using System.Collections.Generic;
using System.Collections;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;

namespace CSharpDesignPattern
{
    public enum Color
    {
        Red, Green, Blue
    }
    public enum Size
    {
        Small, Medium, Large
    }

    public class Product
    {
        public string Name;
        public Color Color;
        public Size Size;
    
        public Product(string name, Color color, Size size)
        {
            if(name == null)
            {
                throw new ArgumentNullException(paramName: nameof(name));
            }
            Name = name;
            Color = color;
            Size = size;
        }
    }

    // wrong
    public class ProductFilter
    {
        public IEnumerable<Product> FilterBySize(IEnumerable<Product> products, Size size)
        {
            foreach (var p in products)
                if (p.Size == size)
                    yield return p;
        }

        public IEnumerable<Product> FilterByColor(IEnumerable<Product> products, Color color)
        {
            foreach (var p in products)
                if (p.Color == color)
                    yield return p;
        }

        public IEnumerable<Product> FilterByColorAndSize(IEnumerable<Product> products, Color color, Size size)
        {
            foreach (var p in products)
                if (p.Color == color && p.Size == size)
                    yield return p;
        }
    }

    public interface ISpecification<T>
    {
        bool IsSatisfied(T t);
    }

    public interface IFilter<T>
    {
        IEnumerable<T> FilterBy(IEnumerable<T> items, ISpecification<T> spec);
    }


    public class ColorSpecification : ISpecification<Product>
    {
        Color color;
        public ColorSpecification(Color color)
        {
            this.color = color;
        }

        public bool IsSatisfied(Product t)
        {
            return this.color == t.Color;
        }
    }

    public class SizeSpecification : ISpecification<Product>
    {
        Size size;
        public SizeSpecification(Size size)
        {
            this.size = size;
        }

        public bool IsSatisfied(Product t)
        {
            return this.size == t.Size;
        }
    }

    public class AndSpecification<T> : ISpecification<T>
    {
        ISpecification<T> first;
        ISpecification<T> second;

        public AndSpecification(ISpecification<T> first,ISpecification<T> second)
        {
            this.first = first;
            this.second = second;
        }
        public bool IsSatisfied(T t)
        {
            return first.IsSatisfied(t) && second.IsSatisfied(t);
        }
    }

    public class BetterFilter : IFilter<Product>
    {
        public IEnumerable<Product> FilterBy(IEnumerable<Product> items, ISpecification<Product> spec)
        {
            foreach (var item in items)
                if (spec.IsSatisfied(item))
                    yield return item;
        }
    }




    class SOLID_OpenClosed
    {
        static void Main(string[] args)
        {
            var apple = new Product("Apple", Color.Green, Size.Small);
            var tree= new Product("Tree", Color.Green, Size.Large);
            var house = new Product("House", Color.Blue, Size.Large);

            Product[] products = { apple, tree, house };

            var pf = new ProductFilter();
            Console.WriteLine("Green products (old):");
            foreach(var p in pf.FilterByColor(products, Color.Green))
            {
                Console.WriteLine($" - {p.Name} is green");
            }


            var bf = new BetterFilter();
            Console.WriteLine("Green products (new):");
            foreach (var p in bf.FilterBy(products, new ColorSpecification(Color.Green)))
            {
                Console.WriteLine($" - {p.Name} is green");
            }


            var bf2 = new BetterFilter();
            Console.WriteLine("Large Blue products (new):");
            foreach (var p in bf2.FilterBy(
                products, 
                new AndSpecification<Product>(
                    new ColorSpecification(Color.Blue), 
                    new SizeSpecification(Size.Large))))
            {
                Console.WriteLine($" - {p.Name} is large and blue");
            }
        }
    }
}

#endif
